gandalf/src/test/java/io/github/btkelly/gandalf/GandalfTest.java

```diff
@@ -0,0 +1,340 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import io.github.btkelly.gandalf.models.Alert;
+import io.github.btkelly.gandalf.models.OptionalUpdate;
+import io.github.btkelly.gandalf.models.RequiredUpdate;
+import okhttp3.Dispatcher;
+import okhttp3.HttpUrl;
+import okhttp3.OkHttpClient;
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
+
+import static org.junit.Assert.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyInt;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.ArgumentMatchers.argThat;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.mockito.MockitoAnnotations.initMocks;
+
+@SuppressWarnings({
+        "RV_RETURN_VALUE_IGNORED",
+        "magicnumber",
+        "TooManyStaticImports"
+})
+public class GandalfTest {
+
+    private static final String MOCK_PACKAGE = "mock.package";
+    public static final String VERSION_NAME = "1.2.3";
+    public static final String ALERT_MESSAGE = "Alert Message";
+    public static final String LATCH_TIMEOUT_MESSAGE = "Latch timeout";
+
+    private MockWebServer mockWebServer;
+
+    @Mock
+    private Context mockContext;
+    @Mock
+    private PackageManager mockPackageManager;
+    @Mock
+    private SharedPreferences mockSharedPreferences;
+
+    @Before
+    public void init() throws IOException, PackageManager.NameNotFoundException {
+        initMocks(this);
+
+        when(mockContext.getPackageName()).thenReturn(MOCK_PACKAGE);
+        when(mockContext.getPackageManager()).thenReturn(mockPackageManager);
+        when(mockContext.getApplicationContext()).thenReturn(mockContext);
+        when(mockContext.getSharedPreferences(anyString(), anyInt())).thenReturn(mockSharedPreferences);
+
+        this.mockWebServer = new MockWebServer();
+        this.mockWebServer.start();
+    }
+
+    @After
+    public void cleanUp() throws IOException, NoSuchFieldException, IllegalAccessException {
+        this.mockWebServer.shutdown();
+        this.mockWebServer = null;
+
+        Field gandalfInstance = Gandalf.class.getDeclaredField("gandalfInstance");
+        gandalfInstance.setAccessible(true);
+        gandalfInstance.set(null, null);
+    }
+
+    @Test
+    public void invalidVersionCodeSafelyCallsNoActionRequired() throws IOException, PackageManager.NameNotFoundException, InterruptedException {
+
+        enqueueMockResponse("invalidVersion.json");
+
+        final CountDownLatch countDownLatch = new CountDownLatch(1);
+
+        GandalfCallback mockCallback = getGandalfCallback(countDownLatch);
+        getGandalfInstance(VERSION_NAME, 6).shallIPass(mockContext, mockCallback);
+
+        boolean status = countDownLatch.await(5, TimeUnit.SECONDS);
+        assertTrue(LATCH_TIMEOUT_MESSAGE, status);
```

Comment:
What's the latch await doing for us here?

Reply:
This is here to allow the async thread to complete before running the verify logic on the mock. I also tried injecting direct thread executors, which worked for success but for some reason caused an issue with OkHttp where it would swallow the exceptions and throw the same OkHttp error every time. This was not ideal for the unit tests so I refactored to use the latch. It's not ideal but it works 🤷‍♂️

Reply:
👍 hopefully none of them take a full 5 seconds but not many tests total so this shouldn't be obnoxious!